File: haskell/2025/day11b/src/Main.hs
module Main ( main ) where

import Control.DeepSeq
import Control.Monad.State.Strict
import Data.Map.Strict ( Map )
import qualified Data.Map.Strict as Map
import System.Clock
import System.Environment
import System.Exit
import System.IO

type Graph = Map String [String]
type MemoKey = (String, Bool, Bool)
type Memo = Map MemoKey Integer

parseLine :: String -> (String, [String])
parseLine line =
    let (k, _:rest) = span (/= ':') line
        vs          = words rest
    in (k, vs)

countPaths :: Graph -> String -> State Memo Integer
countPaths graph start = go start False False
  where
    go :: String -> Bool -> Bool -> State Memo Integer
    go node hasDac hasFft
        | node == "out" = return (if hasDac && hasFft then 1 else 0)
        | otherwise = do
            let hasDac' = hasDac || node == "dac"
                hasFft' = hasFft || node == "fft"
                key = (node, hasDac', hasFft')

            memo <- get
            case Map.lookup key memo of
                Just v  -> return v
                Nothing -> do
                    let neighbors = Map.findWithDefault [] node graph
                    vals <- mapM (\n -> go n hasDac' hasFft') neighbors
                    let total = sum vals
                    modify' (Map.insert key total)
                    return total

process :: String -> Integer
process content =
    let graph = Map.fromList (map parseLine $ filter (not . null) $ lines content)
    in evalState (countPaths graph "svr") Map.empty

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname

File: rust/2025/day11b/src/main.rs
use std::collections::HashMap;
use std::env;
use std::fs;

type Graph = HashMap<String, Vec<String>>;
type MemoKey = (String, bool, bool);
type Memo = HashMap<MemoKey, i64>;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    std::process::exit(1);
}

fn parse_line(line: &str) -> Option<(String, Vec<String>)> {
    let (key, rest) = line.split_once(':')?;
    let neighbors: Vec<String> = rest.split_whitespace().map(String::from).collect();
    Some((key.to_string(), neighbors))
}

fn count_paths(graph: &Graph, node: &str, has_dac: bool, has_fft: bool, memo: &mut Memo) -> i64 {
    if node == "out" {
        return if has_dac && has_fft { 1 } else { 0 };
    }
    let has_dac = has_dac || node == "dac";
    let has_fft = has_fft || node == "fft";
    let key = (node.to_string(), has_dac, has_fft);
    if let Some(&cached) = memo.get(&key) {
        return cached;
    }
    let neighbors = match graph.get(node) {
        Some(ns) => ns.clone(),
        None => Vec::new(),
    };
    let total: i64 = neighbors
        .iter()
        .map(|n| count_paths(graph, n, has_dac, has_fft, memo))
        .sum();
    memo.insert(key, total);
    total
}

fn process(content: &str) -> i64 {
    let graph: Graph = content
        .lines()
        .filter(|line| !line.is_empty())
        .filter_map(parse_line)
        .collect();
    let mut memo = Memo::new();
    count_paths(&graph, "svr", false, false, &mut memo)
}

fn display_duration(duration: u128) {
    print!("elapsed time: ");
    let duration_f = duration as f64;
    if duration_f < 1_000.0 {
        println!("{} ns", duration_f);
    } else if duration_f < 1_000_000.0 {
        println!("{} μs", duration_f / 1_000.0);
    } else if duration_f < 1_000_000_000.0 {
        println!("{} ms", duration_f / 1_000_000.0);
    } else {
        println!("{} s", duration_f / 1_000_000_000.0);
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    let start = std::time::Instant::now();
    match fs::read_to_string(filename) {
        Ok(content) => {
            let result = process(&content);
            let duration = start.elapsed().as_nanos();
            println!("result = {result}");
            display_duration(duration);
        }
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            std::process::exit(1);
        }
    }
}
